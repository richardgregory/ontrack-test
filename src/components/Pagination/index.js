import React from "react";
import PropTypes from "prop-types";
import { LinkContainer } from "react-router-bootstrap";
import { Pagination as BootstrapPagination } from "react-bootstrap";

export const Pagination = ({ className, currentPath, page, totalPages }) => {
  const limit = 2;

  const renderPageNumbers = () => {
    const pages = [];
    const start = page > limit ? page - limit : 1;
    const end = page + limit < totalPages ? page + limit : totalPages;

    for (let i = start; i <= end; i++) {
      pages.push(
        <QueryNavLink
          key={i}
          to={{ pathname: currentPath, search: `?page=${i}` }}
        >
          <BootstrapPagination.Item active={i === page}>
            {i}
          </BootstrapPagination.Item>
        </QueryNavLink>
      );
    }

    return pages;
  };

  return (
    <BootstrapPagination
      className={`${className} justify-content-center align-items-center`}
    >
      <QueryNavLink to={{ pathname: currentPath, search: `?page=${page - 1}` }}>
        <BootstrapPagination.Prev disabled={page <= 1}>
          Previous
        </BootstrapPagination.Prev>
      </QueryNavLink>

      {page > limit + 1 && (
        <>
          <QueryNavLink to={{ pathname: currentPath, search: `?page=1` }}>
            <BootstrapPagination.Item>1</BootstrapPagination.Item>
          </QueryNavLink>

          <BootstrapPagination.Ellipsis disabled />
        </>
      )}

      {renderPageNumbers()}

      {page < totalPages - limit && (
        <>
          <BootstrapPagination.Ellipsis disabled />
          <QueryNavLink
            to={{ pathname: currentPath, search: `?page=${totalPages}` }}
          >
            <BootstrapPagination.Item>{totalPages}</BootstrapPagination.Item>
          </QueryNavLink>
        </>
      )}

      <QueryNavLink to={{ pathname: currentPath, search: `?page=${page + 1}` }}>
        <BootstrapPagination.Next disabled={page >= totalPages}>
          Next
        </BootstrapPagination.Next>
      </QueryNavLink>
    </BootstrapPagination>
  );
};

// Setting isActive to false on react router link
// so active state is managed by bootstrap
const QueryNavLink = ({ children, ...props }) => (
  <LinkContainer isActive={() => false} {...props}>
    {children}
  </LinkContainer>
);

Pagination.propTypes = {
  className: PropTypes.string,
  page: PropTypes.number,
  totalPages: PropTypes.number,
  currentPath: PropTypes.string,
};

export default Pagination;
