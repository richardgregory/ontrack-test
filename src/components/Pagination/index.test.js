import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Pagination from ".";

describe("Pagination", () => {
  it("Should render previous as disabled", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination page={1} totalPages={10} currentPath="/" />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("Should render pages in between", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination page={4} totalPages={10} currentPath="/" />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("Should render next as disabled", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination page={10} totalPages={10} currentPath="/" />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("Should render with no ellipsis at start when page less than 3", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination page={3} totalPages={10} currentPath="/" />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it("Should render with no ellipsis at end when page greater than 7", () => {
    const { container } = render(
      <BrowserRouter>
        <Pagination page={8} totalPages={10} currentPath="/" />
      </BrowserRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
