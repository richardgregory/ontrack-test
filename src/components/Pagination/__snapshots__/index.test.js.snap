// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should render next as disabled 1`] = `
<div>
  <ul
    class="undefined justify-content-center align-items-center pagination"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=9"
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          class="sr-only"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=1"
      >
        1
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=8"
      >
        8
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=9"
      >
        9
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        href="/?page=10"
      >
        10
        <span
          class="sr-only"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
        href="/?page=11"
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          class="sr-only"
        >
          Next
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`Pagination Should render pages in between 1`] = `
<div>
  <ul
    class="undefined justify-content-center align-items-center pagination"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=3"
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          class="sr-only"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=1"
      >
        1
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=2"
      >
        2
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=3"
      >
        3
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        href="/?page=4"
      >
        4
        <span
          class="sr-only"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=5"
      >
        5
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=6"
      >
        6
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=10"
      >
        10
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=5"
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          class="sr-only"
        >
          Next
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pagination Should render previous as disabled 1`] = `
<div>
  <ul
    class="undefined justify-content-center align-items-center pagination"
  >
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
        href="/?page=0"
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          class="sr-only"
        >
          Previous
        </span>
      </span>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        href="/?page=1"
      >
        1
        <span
          class="sr-only"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=2"
      >
        2
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=3"
      >
        3
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=10"
      >
        10
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=2"
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          class="sr-only"
        >
          Next
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pagination Should render with no ellipsis at end when page greater than 7 1`] = `
<div>
  <ul
    class="undefined justify-content-center align-items-center pagination"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=7"
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          class="sr-only"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=1"
      >
        1
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=6"
      >
        6
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=7"
      >
        7
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        href="/?page=8"
      >
        8
        <span
          class="sr-only"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=9"
      >
        9
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=10"
      >
        10
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=9"
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          class="sr-only"
        >
          Next
        </span>
      </a>
    </li>
  </ul>
</div>
`;

exports[`Pagination Should render with no ellipsis at start when page less than 3 1`] = `
<div>
  <ul
    class="undefined justify-content-center align-items-center pagination"
  >
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=2"
      >
        <span
          aria-hidden="true"
        >
          Previous
        </span>
        <span
          class="sr-only"
        >
          Previous
        </span>
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=1"
      >
        1
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=2"
      >
        2
      </a>
    </li>
    <li
      class="page-item active"
    >
      <span
        class="page-link"
        href="/?page=3"
      >
        3
        <span
          class="sr-only"
        >
          (current)
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=4"
      >
        4
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=5"
      >
        5
      </a>
    </li>
    <li
      class="page-item disabled"
    >
      <span
        class="page-link"
        disabled=""
      >
        <span
          aria-hidden="true"
        >
          …
        </span>
        <span
          class="sr-only"
        >
          More
        </span>
      </span>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=10"
      >
        10
      </a>
    </li>
    <li
      class="page-item"
    >
      <a
        class="page-link"
        href="/?page=4"
      >
        <span
          aria-hidden="true"
        >
          Next
        </span>
        <span
          class="sr-only"
        >
          Next
        </span>
      </a>
    </li>
  </ul>
</div>
`;
